File: testdata/jennies/builders/factories/PythonBuilder/builders/promql.py
import typing
from ..cog import builder as cogbuilder
from ..models import promql


class FuncCallExpr(cogbuilder.Builder[promql.FuncCallExpr]):
    _internal: promql.FuncCallExpr

    def __init__(self):
        self._internal = promql.FuncCallExpr()        
        self._internal.type_val = "funcCallExpr"

    def build(self) -> promql.FuncCallExpr:
        """
        Builds the object.
        """
        return self._internal    
    
    def function(self, function: str) -> typing.Self:    
        if not len(function) >= 1:
            raise ValueError("len(function) must be >= 1")
        self._internal.function = function
    
        return self
    
    def args(self, args: list[cogbuilder.Builder[promql.Expr]]) -> typing.Self:    
        args_resources = [r1.build() for r1 in args]
        self._internal.args = args_resources
    
        return self
    
    def arg(self, arg: cogbuilder.Builder[promql.Expr]) -> typing.Self:    
        """
        Modified by veneer 'Duplicate[args]'
        Modified by veneer 'ArrayToAppend'
        """
            
        if self._internal.args is None:
            self._internal.args = []
        
        arg_resource = arg.build()
        self._internal.args.append(arg_resource)
    
        return self
    

"""
Returns the input vector with all sample values converted to their absolute value.
See https://prometheus.io/docs/prometheus/latest/querying/functions/#abs
"""
def abs(v: cogbuilder.Builder[promql.Expr]):
    builder = FuncCallExpr()
    builder.function("abs")
    builder.arg(v)

    return builder

"""
Returns an empty vector if the vector passed to it has any elements (floats or native histograms) and a 1-element vector with the value 1 if the vector passed to it has no elements.
This is useful for alerting on when no time series exist for a given metric name and label combination.
See https://prometheus.io/docs/prometheus/latest/querying/functions/#absent
"""
def absent(v: cogbuilder.Builder[promql.Expr]):
    builder = FuncCallExpr()
    builder.function("absent")
    builder.arg(v)

    return builder

"""
Returns pi.
See https://prometheus.io/docs/prometheus/latest/querying/functions/#trigonometric-functions
"""
def pi():
    builder = FuncCallExpr()
    builder.function("pi")

    return builder

"""
Calculates the φ-quantile (0 ≤ φ ≤ 1) of the values in the specified interval.
See https://prometheus.io/docs/prometheus/latest/querying/functions/#aggregation_over_time
"""
def quantile_over_time(phi: float, v: cogbuilder.Builder[promql.Expr]):
    builder = FuncCallExpr()
    builder.function("quantile_over_time")
    builder.arg(n(phi))
    builder.arg(v)

    return builder

File: testdata/jennies/builders/factories/RustBuilder/src/builders/promql.rs
use crate::cog;
use crate::types::promql::Expr;
use crate::types::promql::FuncCallExpr;

#[derive(Debug, Clone)]
pub struct FuncCallExprBuilder {
    internal: FuncCallExpr,
    errors: Vec<cog::BuildError>,
}

impl FuncCallExprBuilder {
    pub fn new() -> Self {
        let mut builder = Self {
            internal: FuncCallExpr::default(),
            errors: Vec::new(),
        };
        builder.internal.r#type = "funcCallExpr".to_string();

        builder
    }
}

impl Default for FuncCallExprBuilder {
    fn default() -> Self {
        Self::new()
    }
}

impl FuncCallExprBuilder {
    pub fn function(mut self, function: String) -> Self {
        if function.chars().count() < 1 {
            self.errors.push(cog::BuildError::new(
                "function",
                "function length must be >= 1".to_string(),
            ));
        }
        self.internal.function = function;

        self
    }
}

impl FuncCallExprBuilder {
    pub fn args(mut self, args: Vec<impl cog::Builder<Expr>>) -> Self {
        let mut built0 = Vec::new();
        for item0 in args {
            let built1 = match item0.build() {
                Ok(val) => val,
                Err(mut err) => {
                    self.errors.append(&mut err);
                    return self;
                }
            };
            built0.push(built1);
        }
        self.internal.args = built0;

        self
    }
}

/// Modified by veneer 'Duplicate[args]'
/// Modified by veneer 'ArrayToAppend'
impl FuncCallExprBuilder {
    pub fn arg(mut self, arg: impl cog::Builder<Expr>) -> Self {
        let built0 = match arg.build() {
            Ok(val) => val,
            Err(mut err) => {
                self.errors.append(&mut err);
                return self;
            }
        };
        self.internal.args.push(built0);

        self
    }
}

/// Returns the input vector with all sample values converted to their absolute value.
/// See https://prometheus.io/docs/prometheus/latest/querying/functions/#abs
impl FuncCallExprBuilder {
    pub fn abs(v: impl cog::Builder<Expr>) -> Self {
        Self::new().function("abs".to_string()).arg(v)
    }
}

/// Returns an empty vector if the vector passed to it has any elements (floats or native histograms) and a 1-element vector with the value 1 if the vector passed to it has no elements.
/// This is useful for alerting on when no time series exist for a given metric name and label combination.
/// See https://prometheus.io/docs/prometheus/latest/querying/functions/#absent
impl FuncCallExprBuilder {
    pub fn absent(v: impl cog::Builder<Expr>) -> Self {
        Self::new().function("absent".to_string()).arg(v)
    }
}

/// Returns pi.
/// See https://prometheus.io/docs/prometheus/latest/querying/functions/#trigonometric-functions
impl FuncCallExprBuilder {
    pub fn pi() -> Self {
        Self::new().function("pi".to_string())
    }
}

/// Calculates the φ-quantile (0 ≤ φ ≤ 1) of the values in the specified interval.
/// See https://prometheus.io/docs/prometheus/latest/querying/functions/#aggregation_over_time
impl FuncCallExprBuilder {
    pub fn quantile_over_time(phi: f64, v: impl cog::Builder<Expr>) -> Self {
        Self::new()
            .function("quantile_over_time".to_string())
            .arg(NumberLiteralExprBuilder::n(phi))
            .arg(v)
    }
}

impl cog::Builder<FuncCallExpr> for FuncCallExprBuilder {
    fn build(&self) -> Result<FuncCallExpr, Vec<cog::BuildError>> {
        if !self.errors.is_empty() {
            return Err(self.errors.clone());
        }

        Ok(self.internal.clone())
    }
}
